sdk/eventhub/azure-eventhubs/azure/eventhub/_eventprocessor/event_processor.py

```diff
@@ -0,0 +1,314 @@
+    SHUTDOWN: It is explicitly required to stop, this would happen when the EventHubConsumerClient is closed.
+    OWNERSHIP_LOST: It loses the ownership of a partition, this would happend when other EventHubConsumerClient
+    instance claims ownership of the partition.
+    """
+    SHUTDOWN = 0
+    OWNERSHIP_LOST = 1
+
+
+class EventProcessor(object):  # pylint:disable=too-many-instance-attributes
+    """
+    An EventProcessor constantly receives events from one or multiple partitions of the Event Hub
+    in the context of a given consumer group.
+
+    """
+    def __init__(self, eventhub_client, consumer_group_name, on_event, **kwargs):
+        self._consumer_group_name = consumer_group_name
+        self._eventhub_client = eventhub_client
+        self._namespace = eventhub_client._address.hostname  # pylint: disable=protected-access
+        self._eventhub_name = eventhub_client.eh_name
+        self._event_handler = on_event
+        self._partition_id = kwargs.get("partition_id", None)
+        self._error_handler = kwargs.get("on_error", None)
+        self._partition_initialize_handler = kwargs.get("on_partition_initialize", None)
+        self._partition_close_handler = kwargs.get("on_partition_close", None)
+        self._partition_manager = kwargs.get("partition_manager", None)
+        self._initial_event_position = kwargs.get("initial_event_position", EventPosition("-1"))
+
+        self._polling_interval = kwargs.get("polling_interval", 10.0)
+        self._ownership_timeout = self._polling_interval * 2
+
+        self._partition_contexts = {}
+
+        # Receive parameters
+        self._owner_level = kwargs.get("owner_level", None)
+        self._prefetch = kwargs.get("prefetch", None)
+        self._track_last_enqueued_event_properties = kwargs.get("track_last_enqueued_event_properties", False)
+        self._last_enqueued_event_properties = {}
+        self._id = str(uuid.uuid4())
+        self._running = False
+        self._lock = threading.RLock()
+
+        # Each partition consumer is working in its own thread
+        self._working_threads = {}  # type: Dict[str, threading.Thread]
+        self._threads_stop_flags = {}  # type: Dict[str, bool]
+
+        self._callback_queue = queue.Queue(maxsize=10000)  # Right now the limitation of receiving speed is ~10k
+
+    def __repr__(self):
+        return 'EventProcessor: id {}'.format(self._id)
+
+    def _cancel_tasks_for_partitions(self, to_cancel_partitions):
+        to_stop_partition_threads = []
+        with self._lock:
+            for partition_id in to_cancel_partitions:
+                if partition_id in self._working_threads:
+                    thread = self._working_threads.pop(partition_id)
+                    self._threads_stop_flags[partition_id] = True
+                    to_stop_partition_threads.append((partition_id, thread))
+            for partition_id, thread in to_stop_partition_threads:
+                thread.join()
```

Comment:
Is this going to be affected by the fact that we still have the lock here?

Reply:
we have removed this join() and improved the code that ensure every thread would end